File: 01_two_pointers/03_container_with_most_water/container_with_most_water.cpp
#include <vector>
#include <cstddef>

class Solution
{
  public:
    int maxArea(std::vector<int>& height)
    {
        size_t l = 0;
        size_t r = height.size() - 1;

        int max_area = 0;
        while (l < r)
        {
            int area = (r - l) * std::min(height[r], height[l]);
            if (area > max_area)
                max_area = area;
            if (height[r] < height[l])
                --r;
            else
                ++l;
        }
        return max_area;
    }
};

#include <iostream>

int main()
{
    int ary[] = {1, 8, 6, 2, 5, 4, 8, 3, 7};

    std::vector<int> vec;
    for (size_t i = 0; i < sizeof(ary) / sizeof(ary[0]); ++i)
        vec.push_back(ary[i]);

    Solution solution;
    std::cout << solution.maxArea(vec) << "\n";
}

File: NeetCode/01_two_pointers/03_container_with_most_water/container_with_most_water.py
from typing import List

class Solution:
    def maxArea(self, height: List[int]) -> int:
        l = 0
        r = len(height) - 1
        max_area = 0
        while l < r:
            area = (r - l) * min(height[l], height[r])
            if area > max_area:
                max_area = area
            if height[l] > height[r]:
                r -= 1
            else:
                l += 1
        return max_area
        

if __name__ == "__main__":
    solution = Solution()
    print(solution.maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]))
